drivers/ethernet/eth_stm32_hal_priv.h

```diff
@@ -9,6 +9,12 @@
 #include <kernel.h>
 #include <zephyr/types.h>
 
+#if defined(CONFIG_SOC_SERIES_STM32H7X) || defined(CONFIG_SOC_SERIES_STM32F4X)
+#define ETH_HAL_IS_NOT_LEGACY 1
+#else
+#define ETH_HAL_IS_NOT_LEGACY 0
+#endif
```

Comment:
A new Kconfig option to enable the change progressively would be preferable.
It would allow to easily compare both implementations tests results on similar code base.
Also, it would feel less risky to merge either default n or default y (with the possibility, if issues are reported, to disable the time issues are fixed).